heat/core/linalg/svdtools.py

```diff
@@ -0,0 +1,512 @@
+        k = 0
+        counter = 0
+        while k < len(active_nodes):
+            current_idx = active_nodes[k]
+            if used_budget + dims_global[current_idx] > maxmergedim or counter == no_of_merges:
+                current_future_node = current_idx
+                future_nodes.append(current_future_node)
+                used_budget = dims_global[current_idx]
+                counter = 1
+            else:
+                if not used_budget == 0:
+                    send_to[current_idx] = current_future_node
+                used_budget += dims_global[current_idx]
+                counter += 1
+            k += 1
+
+        recv_from = [[]] * no_procs
+        for i in future_nodes:
+            recv_from[i] = [k for k in range(no_procs) if send_to[k] == i]
+
+        if A.comm.rank in future_nodes:
+            # FUTURE NODES
+            # in the future nodes receive local arrays from previous level
+            err_squared_loc = [err_squared_loc] + [
+                torch.zeros_like(err_squared_loc) for i in recv_from[A.comm.rank]
+            ]
+            U_loc = [U_loc] + [
+                torch.zeros(
+                    (A.shape[0], dims_global[i]), dtype=A.larray.dtype, device=A.device.torch_device
+                )
+                for i in recv_from[A.comm.rank]
+            ]
+            for k in range(len(recv_from[A.comm.rank])):
+                A.comm.Recv(U_loc[k + 1], recv_from[A.comm.rank][k], tag=recv_from[A.comm.rank][k])
+                A.comm.Recv(
+                    err_squared_loc[k + 1],
+                    recv_from[A.comm.rank][k],
+                    tag=2 * no_procs + recv_from[A.comm.rank][k],
+                )
+            # concatenate the received arrays
+            U_loc = torch.hstack(U_loc)
+            err_squared_loc = sum(err_squared_loc)
+            level += 1
+            if A.comm.rank == 0 and not silent:
+                print(
+                    "hSVD level %d...\t" % level,
+                    "processes ",
+                    "\t\t".join(["%d" % fn for fn in future_nodes]),
+                )
+            # compute "local" SVDs on the current level
+
+            if len(future_nodes) == 1:
+                safetyshift = 0
+            U_loc, sigma_loc, err_squared_loc_new = compute_local_truncated_svd(
+                level, A.comm.rank, U_loc, maxrank, loctol, safetyshift
+            )
+
+            if len(future_nodes) > 1:
+                # prepare next level or...
+                U_loc = torch.linalg.matmul(U_loc, torch.diag(sigma_loc))
+            err_squared_loc += err_squared_loc_new
+        elif A.comm.rank in active_nodes and A.comm.rank not in future_nodes:
+            # NOT FUTURE NODES
+            # in these nodes we only send the local arrays to the respective future node
+            A.comm.Send(U_loc, send_to[A.comm.rank], tag=A.comm.rank)
+            A.comm.Send(err_squared_loc, send_to[A.comm.rank], tag=2 * no_procs + A.comm.rank)
+
```

Comment:
would it make sense to store `err_squared_loc` as an extra column (or row, whatever) of `U_loc`, and cut down the communication by half?  

This would apply to the two Sends here, the two Recv at line 383-384, and the two Ibcast at lines 426-428 (might be more and I missed it).

Reply:
`err_squared_loc` is a number only, so it cannot be added to `U_loc` as additional row or column without memory overhead (proportional either up to `maxrank` or `A.shape[0]`)

Therefore, I do not know whether the suggested change would really save time.